astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -181,6 +183,38 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> Sequence[Us
         return user_defined_functions
 
 
+@dataclasses.dataclass
+class RetrieveKeeperMapTableDataStep(Step[Sequence[KeeperMapTable]]):
+    zookeeper_client: ZooKeeperClient
+    keeper_map_path_prefix: str | None
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> Sequence[KeeperMapTable]:
+        if self.keeper_map_path_prefix is None:
+            return []
+
+        async with self.zookeeper_client.connect() as connection:
+            change_watch = ChangeWatch()
+            try:
+                children = await connection.get_children(self.keeper_map_path_prefix, watch=change_watch)
+            except NoNodeError:
+                # The path doesn't exist, no keeper map tables to retrieve
+                return []
+
+            raw_tables = {}
+            for child in children:
+                keeper_map_table_path = os.path.join(self.keeper_map_path_prefix, child)
+                data_path = os.path.join(keeper_map_table_path, "data")
+                raw_tables[child] = await connection.get_children_with_data(data_path)
+            tables = [
+                KeeperMapTable(name=table_path, data=[KeeperMapRow(key=k, value=v) for k, v in sorted(table_data.items())])
+                for table_path, table_data in sorted(raw_tables.items())
+            ]
+
+            if change_watch.has_changed:
+                raise TransientException("Concurrent table addition / deletion during KeeperMap backup")
```

Comment:
2. This is what we want, to only monitor direct children.  So this just checks that no table was added / deleted while the step is running, _not_ that the table data itself has changed.

`get_children_with_data` is what returns a consistent view of the data in the table even if that data is changing.  It could be the case that this function never returns if znodes are constantly being added / removed, so I should probably put more logging around this and include some timeouts.

1. In this case the table contains around 4 rows for every kafka topic that is being synced and is updated after every write from the connector.  But nodes are only being updated, not added, so should complete correctly.

3. `TransientException` should mean the step is retried.  Which isn't great if a table has been removed, since it will keep failing.

Reply:
2. Is it correct that each child is a KeeperMap table and we care about the only child (representing table named connect_state).
Yes, in this case it is cool. (Probably you could even not care about the changes)
1. as far as I can see, is a row per topic per partition, probably there will not be too many rows in total.
and we may decide separately if we allow user create their keeper map tables (I do not think we are ready)